apps/meeting/src/providers/DataMessagesProvider/index.tsx

```diff
@@ -0,0 +1,109 @@
+import { RosterType, useAudioVideo, useMeetingManager, useRosterState } from 'amazon-chime-sdk-component-library-react';
+import { DataMessage } from 'amazon-chime-sdk-js';
+import React, { useEffect, useReducer, createContext, useContext, FC, useCallback, useRef } from 'react';
+import { DataMessagesActionType, initialState, ChatDataMessage, reducer } from './state';
+
+interface DataMessagesStateContextType {
+  sendMessage: (message: string) => void;
+  messages: ChatDataMessage[];
+}
+
+const DataMessagesStateContext = createContext<DataMessagesStateContextType | undefined>(undefined);
+
+export const DataMessagesProvider: FC = ({ children }) => {
+  const meetingManager = useMeetingManager();
+  const audioVideo = useAudioVideo();
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const { roster } = useRosterState();
+  const rosterRef = useRef<RosterType>({});
+
+  useEffect(() => {
+    if (!audioVideo) {
+      return;
+    }
+    audioVideo.realtimeSubscribeToReceiveDataMessage('ChimeComponentLibraryDataMessage', handler);
+    return () => {
+      audioVideo.realtimeUnsubscribeFromReceiveDataMessage('ChimeComponentLibraryDataMessage');
+    };
+  }, [audioVideo]);
+
+  useEffect(() => {
+    if (roster) {
+      rosterRef.current = { ...rosterRef.current, ...roster };
+    }
+  }, [roster]);
+
+  const handler = useCallback(
+    (dataMessage: DataMessage) => {
+      if (!dataMessage.throttled) {
+        const isSelf =
+          dataMessage.senderAttendeeId === meetingManager.meetingSession?.configuration.credentials?.attendeeId;
+        if (isSelf) {
+          dispatch({
+            type: DataMessagesActionType.ADD,
+            payload: {
+              message: new TextDecoder().decode(dataMessage.data),
```

Comment:
For understanding, why we need to handle `dataMessage.data` differently than line 57.

```js
message: new TextDecoder().decode(dataMessage.data),
message: data.message,
```

Reply:
So, when a local attendee sends the data message, it is sent over the peer-connection for remotes + triggered to self so that the the subscribed data message handler gets all updates in one place. This is the same logic in JS SDK as well. Basically, update the state only when the data message has successfully received from local attendee or remote.

Reply:
For a local attendee, the data message received from the remote is a JSON, but when sending from local to remote, the new DataMessage creation accepts only Uint8Array and not a string hence had to handle separate cases.

Reply:
Got it! Thanks for your detailed explanation. 👍
Learned something new :D.